bpf/tests/pktgen.h

```diff
@@ -138,6 +164,29 @@ int pktgen__free_layer(const struct pktgen *builder)
 	return -1;
 }
 
+static __always_inline
+__attribute__((warn_unused_result))
+int ctx_data_valid(__maybe_unused struct __ctx_buff *ctx, __maybe_unused __u16 off)
+{
+	int ret = 0;
+
+	/* try to open up a range on the pkt reg */
+	asm volatile("r1 = *(u32 *)(%[ctx] +0)\n\t"
+		     "r2 = *(u32 *)(%[ctx] +4)\n\t"
+		     "%[off] &= %[offmax]\n\t"
+		     "r1 += %[off]\n\t"
+		     "if r1 > r2 goto +2\n\t"
+		     "%[ret] = 0\n\t"
+		     "goto +1\n\t"
+		     "%[ret] = %[errno]\n\t"
+		     : [ret]"=r"(ret)
+		     : [ctx]"r"(ctx), [off]"r"(off),
+		       [offmax]"i"(0xff), [errno]"i"(-EINVAL)
+		     : "r1", "r2");
+
+	return ret;
+}
+
```

Comment:
I am not 100% about this function. I believe this is masking a bigger issue we recently discovered but have not yet patched.

Long story short. Can you see if you can leave this function out. And instead update the structure of the pktgen like so:

```c
struct pktgen {
	struct __ctx_buff *ctx;
	__u16 cur_off;
	__u64 : 48;
	__u16 layer_offsets[PKT_BUILDER_LAYERS];
	enum pkt_layer layers[PKT_BUILDER_LAYERS];
};
```

The verifier works with 8-byte slots of memory, if you pack store more than 1 variable in a slot, the verifier will simply not record bounds check information. Adding the `__u64 : 48;` ensures that our variable never shares a slot with others on the stack which I think also fixes the behavior you were describing at the callsite in `pktgen__append_ipv6_extension_header`

Reply:
```
 struct pktgen {
        struct __ctx_buff *ctx;
        __u16 cur_off;
+       __u64 : 48;
        __u16 layer_offsets[PKT_BUILDER_LAYERS];
        enum pkt_layer layers[PKT_BUILDER_LAYERS];
 };
@@ -350,7 +351,9 @@ struct ipv6_opt_hdr *pktgen__append_ipv6_extension_header(struct pktgen *builder
         * code that verifier doesn't understand. try to open up a range on
         * the pkt reg
         */
-       if (ctx_data_valid(builder->ctx, builder->cur_off))
+       // if (ctx_data_valid(builder->ctx, builder->cur_off))
+       //      return NULL;
+       if ((void *) hdr + length > ctx_data_end(builder->ctx))
                return NULL;
 
        hdr->hdrlen = hdrlen;
```
```
                ; if (!hdr)
                217: (15) if r1 == 0x0 goto pc+854
                 R0_w=fp-44 R1=pkt(id=8,off=0,r=0,umax_value=65534,var_off=(0x0; 0xffff),s32_max_value=65535,u32_max_value=65535) R2=pkt(id=8,off=8,r=0,umax_value=65534,var_off=(0x0; 0xffff),s32_max_value=65535,u32_max_value=65535) R3_w=invP6 R4_w=inv6 R5_w=fp-50 R6=ctx(id=0,off=0,imm=0) R7=invP(id=4,umax_value=65510,var_off=(0x0; 0xffff)) R8_w=invP(id=0,umin_value=24,umax_value=65534,var_off=(0x0; 0xffff)) R10=fp0 fp-8=????0000 fp-16=0000mmmm fp-24=mmmmmmmm fp-32=mmmm0000 fp-40=mmmmmm00 fp-48=??????mm fp-56=ctx
                ; return (void *)(unsigned long)ctx->data_end;
                218: (61) r3 = *(u32 *)(r6 +4)
                ; if ((void *) hdr + length > ctx_data_end(builder->ctx))
                219: (2d) if r2 > r3 goto pc+852
                 R0=fp-44 R1=pkt(id=8,off=0,r=0,umax_value=65534,var_off=(0x0; 0xffff),s32_max_value=65535,u32_max_value=65535) R2=pkt(id=8,off=8,r=0,umax_value=65534,var_off=(0x0; 0xffff),s32_max_value=65535,u32_max_value=65535) R3=pkt_end(id=0,off=0,imm=0) R4=inv6 R5=fp-50 R6=ctx(id=0,off=0,imm=0) R7=invP(id=4,umax_value=65510,var_off=(0x0; 0xffff)) R8=invP(id=0,umin_value=24,umax_value=65534,var_off=(0x0; 0xffff)) R10=fp0 fp-8=????0000 fp-16=0000mmmm fp-24=mmmmmmmm fp-32=mmmm0000 fp-40=mmmmmm00 fp-48=??????mm fp-56=ctx
                220: (b7) r2 = 0
                ; hdr->hdrlen = hdrlen;
                221: (73) *(u8 *)(r1 +1) = r2
                invalid access to packet, off=1 size=1, R1(id=8,off=1,r=0)
                R1 offset is outside of the packet
                processed 190 insns (limit 1000000) max_states_per_insn 0 total_states 11 peak_states 11 mark_read 9
FAIL
FAIL    github.com/cilium/cilium/test/bpf_tests 2.076s
FAIL
make: *** [Makefile:92: run_bpf_tests] Error 1
make: Leaving directory '/home/ubuntu/cilium/test'
```

Sorry for the late reply.

I tried adding the patch mentioned above, but found that the range of the r1 register is still 0.

Reply:
Alright, then we might still have separate issues, please disregard my comment